File: sample/Dracula.hs
module Main where

-- Once upon a time...
data Vampire = Vampire { location :: String
                       , birthDate :: Int
                       , deathDate :: Int
                       , weaknesses :: [String]
                       }

getAge :: Vampire -> Int
getAge = calcAge

calcAge :: Vampire -> Int
calcAge v = let death = deathDate v
                birth = birthDate v
            in death - birth

main :: IO ()
main = do
  -- ...there was a guy named Vlad
  let vampire = Vampire {
      location = "Transylvania",
      birthDate = 1428,
      deathDate = 1476,
      weaknesses = ["Sunlight", "Garlic"]
  }
  print $ getAge vampire

File: sample/dracula.lua
Vampire = {}

function Vampire:new(obj)
    local obj = obj or {}
    self.__index = self
    setmetatable(obj, self)
    
    return obj
end

function Vampire.calc_age(self)
    return self.death_date - self.birth_date
end

function Vampire:get_age()
    return self:calc_age()
end

local vampire = Vampire:new{
    location = "Transylvania",
    birth_date = 1428,
    death_date = 1476,
    weaknesses = {"Sunlight", "Garlic"}
}

vampire:get_age()
